lib/LanguagePicker.js

```diff
@@ -1,62 +1,67 @@
-    throw new Error('languages parameter must be an array of language codes');
-  }
+function LanguagePicker(lang = 'en', config = {}) {
+  let fallbackMap = {};
+  let fallbacks = [];
 
-  this._nameTag = opts.nameTag;
-  this._multiTag = opts.multiTag;
-  this._languages = {};
+  this.nameTag = config.nameTag;
+  // The prefix is either given or is the nameTag
+  // with an underscore
+  // See Babel.js#24
+  this.prefix = config.multiTag || (this.nameTag && `${this.nameTag}_`) || '';
+  fallbackMap = config.languageMap || {};
 
-  this._priority = 0;
-  this._addLangsArray(opts.languages, this._multiTag);
+  // Add known fallbacks for the language
+  if (fallbackMap[lang]) {
+    fallbacks = Array.isArray(fallbackMap[lang]) ?
+      fallbackMap[lang] : [fallbackMap[lang]];
+  }
 
-  const scripts = ls.adjust({ prefix: this._multiTag, override: dataOverrides });
+  // Use the given language as first choice
+  fallbacks = [lang].concat(fallbacks);
 
-  // Use any language code that hasn't been explicitly listed if it uses the same script
-  const primaryScript = (opts.languages.length > 0 && scripts[this._multiTag + opts.languages[0]]) || 'Latn';
-  this._addLangsArray(Object.keys(scripts).filter(lng => scripts[lng] === primaryScript));
+  // Fallback to English
+  if (fallbacks.indexOf('en') === -1) {
+    fallbacks.push('en');
+  }
 
-  // Use any Latn language if the primary script is not Latn
-  if (primaryScript !== 'Latn') {
-    this._addLangsArray(Object.keys(scripts).filter(lng => scripts[lng] === 'Latn'));
+  // Fallback to nameTag
+  if (this.nameTag) {
+    fallbacks.push(this.nameTag);
   }
 
-  // Use "name" as the last fallback by default
-  this._languages[this._nameTag] = Number.MAX_VALUE;
-}
+  // Add prefix to all languages if exists
+  // eslint-disable-next-line arrow-body-style
+  fallbacks = fallbacks.map((code) => {
+    return code === this.nameTag ? code : this.prefix + code;
+  });
 
-LanguagePicker.prototype._addLangsArray = function _addLangsArray(langs, prefix) {
-  for (const lang of langs) {
-    const id = prefix ? prefix + lang : lang;
-    if (!Object.prototype.hasOwnProperty.call(this._languages, id)) {
-      this._priority += 1;
-      this._languages[id] = this._priority;
-    }
-  }
-};
+  this.fallbacks = fallbacks;
+}
 
 LanguagePicker.prototype.newProcessor = function newProcessor() {
-  const fallbackLanguages = this._languages;
-  let bestValue;
-  let bestIndex;
-  let firstUnk;
+  const { fallbacks, nameTag, prefix } = this;
+  let bestChoiceLang = nameTag || (prefix ? `${prefix}en` : 'en');
+  let bestChoiceValue;
+  let firstChoiceValue;
 
   return {
     addValue(lang, value) {
-      const order = fallbackLanguages[lang];
-      if (order) {
-        if (bestIndex === undefined || bestIndex > order) {
-          bestIndex = order;
-          bestValue = value;
-        }
-      } else if (firstUnk === undefined) {
-        firstUnk = value;
+      if (
+        fallbacks.indexOf(lang) > -1 &&
```

Comment:
perf - searching twice for the `lang`. Use a temp var.

Reply:
Good point; done